allennlp/training/deepspeed/trainer.py

```diff
@@ -0,0 +1,703 @@
+        moving_average: Optional[MovingAverage] = None,
+        batch_callbacks: List[BatchCallback] = None,
+        epoch_callbacks: List[EpochCallback] = None,
+        end_callbacks: List[EpochCallback] = None,
+        trainer_callbacks: List[TrainerCallback] = None,
+        distributed: bool = False,
+        local_rank: int = 0,
+        world_size: int = 1,
+        num_gradient_accumulation_steps: int = 1
+    ) -> None:
+        super().__init__(serialization_dir, cuda_device, distributed, local_rank, world_size)
+        self.model = model
+
+        self.model_engine = deepspeed_engine
+        self.optimizer = deepspeed_engine.optimizer
+
+        self.data_loader = data_loader
+        self._validation_data_loader = validation_data_loader
+
+        if patience is None:  # no early stopping
+            if validation_data_loader is not None:
+                logger.warning(
+                    "You provided a validation dataset but patience was set to None, "
+                    "meaning that early stopping is disabled"
+                )
+        elif (not isinstance(patience, int)) or patience <= 0:
+            raise ConfigurationError(
+                '{} is an invalid value for "patience": it must be a positive integer '
+                "or None (if you want to disable early stopping)".format(patience)
+            )
+
+        # For tracking is_best_so_far and should_stop_early
+        self._metric_tracker = MetricTracker(patience, validation_metric)
+        # Get rid of + or -
+        self._validation_metric = validation_metric[1:]
+
+        self._num_epochs = num_epochs
+
+        if checkpointer is not None:
+            self._checkpointer = checkpointer
+        else:
+            self._checkpointer = DeepspeedCheckpointer(serialization_dir)
+
+        self._moving_average = moving_average
+        self._batch_callbacks = batch_callbacks or []
+        self._epoch_callbacks = epoch_callbacks or []
+        self._end_callbacks = end_callbacks or []
+
+        for callback in trainer_callbacks or []:
+            self._batch_callbacks.append(callback.batch())
+            self._epoch_callbacks.append(callback.epoch())
+            self._end_callbacks.append(callback.end())
+
+        # We keep the total batch number as an instance variable because it
+        # is used inside a closure for the hook which logs activations in
+        # `_enable_activation_logging`.
+        self._batch_num_total = 0
+
+        self._tensorboard = tensorboard_writer or TensorboardWriter(serialization_dir)
+        self._tensorboard.get_batch_num_total = lambda: self._batch_num_total
+        self._tensorboard.enable_activation_logging(self.model)
+
+        self._last_log = 0.0  # time of last logging
+
+        self._num_gradient_accumulation_steps = num_gradient_accumulation_steps
+
+        self._pytorch_model = self.model
+
+    def batch_outputs(self, batch: TensorDict, for_training: bool) -> Dict[str, torch.Tensor]:
```

Comment:
How many methods in here actually differ from `GradientDescentTrainer`. Can you subclass that one and only override the methods that are actually different?

Reply:
To be sure, I don't want to compromise `GradientDescentTrainer` to reduce code duplication here, but if we can reduce it without compromising, let's do it.